extensions/amp-image-slider/0.0/amp-image-slider.js

```diff
@@ -0,0 +1,566 @@
+        'Experiment <amp-image-slider> disabled');
+
+    const children = this.getRealChildren();
+
+    for (let i = 0; i < children.length; i++) {
+      const child = children[i];
+      if (child.hasAttribute('before')) {
+        switch (child.tagName.toLowerCase()) {
+          case 'amp-img':
+            this.leftAmpImage_ = child;
+            break;
+          case 'div':
+            this.leftLabel_ = child;
+            break;
+        }
+      } else if (child.hasAttribute('after')) {
+        switch (child.tagName.toLowerCase()) {
+          case 'amp-img':
+            this.rightAmpImage_ = child;
+            break;
+          case 'div':
+            this.rightLabel_ = child;
+            break;
+        }
+      }
+    }
+
+    if (!this.leftAmpImage_ || !this.rightAmpImage_) {
+      return null;
+    }
+
+    if (!isExperimentOn(this.win, 'layers')) {
+      // When layers not enabled
+      this.setAsOwner(this.leftAmpImage_);
+      this.setAsOwner(this.rightAmpImage_);
+    }
+
+    this.container_.classList.add('i-amphtml-image-slider-container');
+    this.element.appendChild(this.container_);
+
+    this.buildImages();
+    this.buildBar();
+
+    return Promise.resolve();
+  }
+
+  /**
+   * Build image structures
+   */
+  buildImages() {
+    // Hierarchy:
+    // leftMask
+    //   |_ leftAmpImage
+    // rightMask
+    //   |_ rightAmpImage
+    this.container_.appendChild(this.rightMask_);
+    this.container_.appendChild(this.leftMask_);
+
+    this.rightMask_.appendChild(this.rightAmpImage_);
+    this.rightMask_.classList.add('i-amphtml-image-slider-right-mask');
+
+    if (this.rightLabel_) {
+      this.rightLabelWrapper_ = this.doc_.createElement('div');
+      this.rightLabelWrapper_.classList
+          .add('i-amphtml-image-slider-right-label-wrapper');
+      this.rightLabel_.classList.add('i-amphtml-image-slider-right-label');
+      this.rightLabelWrapper_.appendChild(this.rightLabel_);
+      this.rightMask_.appendChild(this.rightLabelWrapper_);
+    }
+    this.leftMask_.appendChild(this.leftAmpImage_);
+
+    this.leftMask_.classList.add('i-amphtml-image-slider-left-mask');
+    this.leftAmpImage_.classList.add('i-amphtml-image-slider-over-image');
+
+    if (this.leftLabel_) {
+      this.leftLabelWrapper_ = this.doc_.createElement('div');
+      this.leftLabelWrapper_.classList
+          .add('i-amphtml-image-slider-left-label-wrapper');
+      this.leftLabel_.classList.add('i-amphtml-image-slider-left-label');
+      this.leftLabelWrapper_.appendChild(this.leftLabel_);
+      this.leftMask_.appendChild(this.leftLabelWrapper_);
+    }
+  }
+
+  /**
+   * Build the image slider bar
+   */
+  buildBar() {
+    if (this.isHoverSlider_) {
+      this.buildHoverSliderBar();
+    } else {
+      this.buildTapSliderBar();
+    }
+  }
+
+  /**
+   * Build slider bar that follows mouse movement
+   */
+  buildHoverSliderBar() {
+    this.container_.appendChild(this.bar_);
```

Comment:
Unless necessary for layout reasons (e.g. `<amp-img>` cannot be without parent), you should try to create elements fully and append them to the DOM last. E.g. You should first append barstick to bar, apply all CSS classes, and append `this.bar_` to `this.container_` last. This is done to trigger the least amount of reflows. (E.g. https://howchoo.com/g/mmu0nguznjg/learn-the-slow-and-fast-way-to-append-elements-to-the-dom)